Fixes Issue1658: Sign blocks using pre-image for noise

hewison-chris opened the pull request:
The signature noise is calculated by adding the enrollment commitment and the pre-image at the block height being signed.

hewison-chris commented:
I need to update this once the https://github.com/bosagora/agora/pull/1857 is merged and a few more ledger tests are re-fixed.

hewison-chris commented:
> I need to update this once the #1857 is merged and a few more ledger tests are re-fixed.

Now all 🟢

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -797,6 +797,53 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    public Hash getPreimage (in Height height) @safe nothrow
```
I don't think you need this function having the parameter, `height` for block signing. Instead, you can use the following two functions. The nominator already has a `KeyPair` object so the second `getPreimage` can be used for that.
The name of `getValidatorPreImageAt` would be better because it already has `getValidatorPreimage` function.

hewison-chris replied:
But if we have the cycle in memory then isn't it more efficient to fetch it from there rather than finding the utxo from our public key and then querying the database?

linked0 replied:
Right, it is. But I think the three overridden functions should do a similar operation, which keeps other developers from confusing the code. If you prefer your way, I would recommend changing the second and third functions to `getValidatorPreimageAt`.

hewison-chris replied:
Updated

linked0 replied:
I would suggest renaming the function again. There are so many `getPreimage`-like functions. It's because the `EM` manages its own enrollment and the validator set. I think we need to make the names of the functions clearer. How about changing it to `getThisValidatorPreimage` or `getThisPreimage`?

hewison-chris replied:
How about `getOurPreimage`?

linked0 replied:
Good!

hewison-chris replied:
I pushed a new commit as this was added via another PR already merged.

linked0 reviewed (commented):
One comment, I will review this more.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -797,6 +797,67 @@ public class EnrollmentManager
+        return this.cycle[height];
+    }
+
+    /***************************************************************************
+
+        Get the pre-image hash for a validator by either the public key or the
+        enrollment key and height
+
+        Params:
+            public_key = validator public key
+            enroll_key = UTXO enrollemnt key
+            height = the block height of pre-image to be returned
+
+        Returns:
+            The preimage hash if found otherwise Hash.init
+
+    ***************************************************************************/
+
+    public Hash getValidatorPreimageAt (in PublicKey public_key, in Height height) @safe nothrow
+    {
+        return this.getValidatorPreimageAt(getEnrollmentForKey(public_key), height);
+    }
+
+    /// Ditto
+    public Hash getValidatorPreimageAt (in Hash enroll_key, in Height height) @safe nothrow
+    {
+        return this.validator_set.getPreimageAt(enroll_key, height).hash;
+    }
```
We already have way too many one liners which forward, let's not add some.
Most of those one liner give a slightly different piece of data, I'd really wish we consolidate them.

hewison-chris replied:
I'll drop the second one as it is only being used here.

hewison-chris committed:
Use preimage to generate the block signature noise

This is to prevent nodes from signing more than one block at a certain
height. The preimage, which is unique per block height, is used to
generate the signature noise. If the same signature noise is used
multiple times then the node's private key can be calculated. There is a
unit test to demonstrate this calculation.

hewison-chris commented:
A lot less changes this time as block signature validation was recently moved up into the Ledger. Thanks to @Geod24

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1228,6 +1238,7 @@ version (unittest)
     private final class TestLedger : ValidatingLedger
     {
         public this (KeyPair key_pair,
+            const KeyPair[] key_pairs = genesis_validator_keys,    // Genesis key pairs
```
Why this ?

hewison-chris replied:
In theory we could use any key pairs but as we always use the test GenesisBlock I could remove this.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -798,8 +798,18 @@ public class Ledger
             const CR = this.enroll_man.getCommitmentNonce(K, block.header.height);  // commited R
             if (CR == Point.init)
+            {
+                log.trace("Block#{}: Validator {} (idx: {}) Couldn't find commitment for this validator",
+                          block.header.height, K, idx);
                 return "Block: Couldn't find commitment for this validator";
-            Point R = CR + challenge.toPoint();
+            }
+            const preimage = this.enroll_man.getValidatorPreimageAt(K, Height(block.header.height - 1)); // Preimage at previous block height
+            if (preimage == Hash.init)
+            {
+                log.trace("Block#{}: Validator {} (idx: {}) missing preimage", block.header.height, K, idx);
+                continue;
+            }
+            Point R = CR + Scalar(preimage).toPoint();
```
This is not correct. If a validator has signed, then its pre-image needs to be present, and we need to be able to find its commitment. Not being able to find the commitment is a bug in the node (I think), but not finding the pre-image (unless there's a bug in the node) indicates that the block is invalid.

hewison-chris replied:
I am not sure if we will always have the pre image when we want to validate the block. I'll replace with returning error and see if it still works in system integration.

Geod24 replied:
Well a validator is either:
- Listed as missing;
- Not listed as missing, nor signing;
- Listed as signing and not listed as missing;

If a validator is listed as missing, and only listed as missing, its pre-image is not taken into account in the computation of the random seed. For the other two cases, it is known and will be taken into account.

hewison-chris replied:
Updated and is still 🟢

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -616,9 +644,8 @@ public class ValidatorSet
-                "AND enrolled_height + distance >= ? AND active = ?",
-                enroll_key, height.value, height.value,
-                EnrollmentStatus.Active);
+                "AND enrolled_height + distance >= ? ORDER BY active DESC",
```
Was looking into this earlier, wouldn't the following work:
```suggestion
                "AND enrolled_height + distance >= ? ORDER BY enrolled_height + distance DESC",
```
?

Which we will later replace with `height` once we finish the `distance` => `height` change.

hewison-chris replied:
Yes that would also work and prepare for the coming change. I'll update it.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -750,21 +756,28 @@ extern(D):
         }
-        log.trace("collectBlockSignature: VALID block signature at height {} for node {}",
-            block_sig.height, block_sig.public_key);
+        log.trace("collectBlockSignature height {} node {}: VALID block signature " ~
+            "- Enrollment commitment (CR): {}, signing commitment (R): {}, preimage: {}",
+                block_sig.height, block_sig.public_key, CR, R, preimage);
```
Can we remove this ?

Geod24 replied:
Or shorten it greatly, to the tune of "Accepted signature" + height + validator key

hewison-chris replied:
Removed

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -750,21 +756,28 @@ extern(D):
         const Scalar block_challenge = block_hash;
         // Fetch the R from enrollment commitment for signing validator
         const CR = this.enroll_man.getCommitmentNonce(block_sig.public_key, block_sig.height);
+        // Get the preimage at ledger height
+        const preimage = this.enroll_man.getValidatorPreimageAt(block_sig.public_key, Height(block_sig.height - 1));
+        if (preimage == Hash.init)
+        {
+            log.warn("No preimage for {} to sign block at height {}", block_sig.public_key, block_sig.height);
+            return false;
+        }
```
If we don't know the pre-image we shouldn't (couldn't) have externalized the block, because we used it for the random seed.

hewison-chris replied:
But we call this also in `SCP confirm` which I think is before we externalize.

Geod24 replied:
The same reasoning applies: The process of creating the block should not have built a block which is not consistent.

hewison-chris replied:
I left this check as this function is called when a block signature is received over the network. We want to ignore the signature if we do not have the `pre-image` otherwise the node could send us a block signature without revealing the `preimage` and cause us problems.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -483,8 +483,9 @@ extern(D):
         {
-            // If it fails signature verification, it might not originate from said key
-            log.trace("Envelope failed signature verification for {}", public_key);
+            log.trace("INVALID Envelope signature {} for public key {} " ~
+                "challenge {}",
+                envelope.signature, public_key, challenge.toString(PrintMode.Clear));
```
?

hewison-chris replied:
It may be an envelope from a bad node so we log at trace level the details. The challenge is a Scalar so we need to print it in the clear.

Geod24 replied:
I mean, we shortened this log message just last PR, and you are essentially reverting those changes here

hewison-chris replied:
Reverted.

Geod24 commented on source/agora/consensus/data/Block.d:
```diff
@@ -775,8 +775,25 @@ unittest
+
     mixin AddLogger!();
 
+    Height[PublicKey] first_enrolled;
+
+    // Note that the height here is for the next block to be externalized
+    Hash getPreimageForValidator (in KeyPair key_pair, in Height height) @safe nothrow
+    {
+        if ((key_pair.address in first_enrolled) is null)
+            first_enrolled[key_pair.address] = height;
+        auto times = 1999 - height - first_enrolled[key_pair.address];
+        log.trace("getPreimageForValidator: First enrollment for {} was at height {}. hashing {} times", key_pair.address, height, times);
+        auto seed = hashMulti(key_pair.secret, "consensus.preimages", 0);
+        foreach (i;  0 .. times)
+            seed = hashFull(seed);
+        return seed;
+    }
```
Why do we need this ?

hewison-chris replied:
It is test code to provide the pre images for the unit tests in `node.Ledger` and `data.Block`.

Geod24 replied:
I think `multiSigTestBlock` needs to go. Regarding pre-image, it depends when they are needed. If they are needed during the initial loading, that's where the pre-image should be propagated. If it's before adding a block in test code, the Ledger should provide a mean for the user (in unittest) to externalize the pre-image for a given validator at a certain distance.

hewison-chris replied:
Removed by merged PR: https://github.com/bosagora/agora/pull/2052

AndrejMitrovic commented on source/agora/consensus/data/Block.d:
```diff
@@ -817,3 +835,54 @@ version (unittest)
     }
 }
+
+/// demonstrate signing two blocks at height 1 to reveal private node key
```
I have a question. What if more than 1 node signs two different blocks at the same height? For example 10 nodes sign block A and 2 nodes (call them X and Y) sign block B.

This would "reveal" the private key of X+Y, but not X's key or Y's key individually right?

hewison-chris replied:
It should be still possible to extract the private key for X and Y by using the two sigs received from each of them.

AndrejMitrovic replied:
But not if the signature is returned summed together in a block, for example in a call to `getBlocksFrom`.

Anyway it's not something that needs to be dealt with in this PR, but I had this thought for a long while now.

hewison-chris replied:
We don't use the signature with multiple signatures combined in this extraction process. Note that issue https://github.com/bosagora/agora/issues/408 covers the actual implementation to extract the keys for real.

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -797,6 +797,59 @@ public class EnrollmentManager
+        Get the pre-image hash for this validator
+
+        Params:
+            enroll_key = UTXO enrollemnt key
```
Remove this, please.

hewison-chris replied:
Removed.

linked0 commented on source/agora/consensus/data/Block.d:
```diff
@@ -796,11 +813,12 @@ version (unittest)
-                cycleForValidator(key.address), i, key.address, rc.toPoint(), s);
+            log.trace("multiSigTestBlock: height: {}, cycle: {} index: {}. Validator {}, R = {}, preimage = {}",
+                block.header.height, cycleForValidator(key.address), i, key.address, rc.toPoint(), preimage);
             sigs ~= Signature(R.V, s);
```
We don't need the variable, `s`.

hewison-chris replied:
I don't follow.

hewison-chris replied:
ok I think I know. s is only used once.

hewison-chris replied:
I also corrected a bug in the logging.

linked0 reviewed (approved):
LGTM with two suggestions.

hewison-chris commented:
Still 🟢

hewison-chris commented:
Green.

Geod24 commented on source/agora/consensus/data/Block.d:
```diff
@@ -775,8 +775,25 @@ unittest
+
     public Block multiSigTestBlock (ref Block block,
         ulong delegate (PublicKey) cycleForValidator,
         KeyPair[] keys) @trusted nothrow
```
This function needs to go.

hewison-chris replied:
Will create a PR to remove this.

hewison-chris replied:
Removed by merged PR: https://github.com/bosagora/agora/pull/2052

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1535,12 +1578,17 @@ unittest
                 new FeeManager(),
                 new MockClock(params.GenesisTimestamp +
                               (blocks.length * params.BlockInterval.total!"seconds")));
+            if (blocks)
+            {
+                blocks[1 .. $].map!(b => b.header.height).each!(h => simulatePreimages(this.log, enroll_man, genesis_validator_keys, h));
+                blocks[1 .. $].each!(b => this.addValidatedBlock(b));
+            }
```
Instead of patching it here, we should feed the `ThrowingLedger` the right blocks (valid ones).

hewison-chris replied:
I don't see how we can store the `pre-images`  without an `EnrollmentManager` which is a member of the `Ledger`. That is why I patch the blocks here in the constructor.

hewison-chris replied:
On second thought I could add `EnrollmentManager` as a parameter to `TestLedger` and `ThrowingLedger`. It would also mean that a few of the existing parameters could be removed as they are members of `EnrollmentManager`.

Geod24 replied:
Why not instantiate it in the ctor, populate the pre-images, and then build the ledger ?
My main concern is using a different road to add the block instead of using the code that is already doing this in the Ledger.

hewison-chris replied:
Do you mean using `acceptBlock` to add blocks rather than `addValidatedBlock` which is what I use here?

hewison-chris replied:
I just tried replacing `addValidatedBlock` with `acceptBlock` and the Ledger tests pass still. That means it is now using the same code to add the blocks. Just inside the constructor here rather than before, which will involve updating a lot of calling code.

Geod24 replied:
I mean that L1581 (previously L1538) should not be modified (`blocks.takeExactly(1)`).

The Ledger has recovery code in the constructor: https://github.com/bosagora/agora/blob/03a97193bc93d0cc80167e0e85892e635d3b62a5/source/agora/node/Ledger.d#L182-L220

What L1581 does is to *completely bypass that code*, then replicate its functionality here. My guess as for why this is being done is that the EM does not have the pre-images. One way to work around this is by adding them ahead of time, which require instantiating the EM, adding the pre-images, then passing it to Ledger's ctor.
However that might currently only work with 1 consensus round, so further changes might be necessary. But *bypassing a piece of code just to duplicate it in unittest* is going to come back to bite us back very soon.

hewison-chris replied:
ok now I see what you mean, thanks for the explanation. I will add the pre-images to the EM in the `TestLedger` and `ThrowingLedger` before calling the parent constructor.

hewison-chris replied:
Has been refactored.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1251,15 +1262,24 @@ version (unittest)
                 new FeeManager(stateDB, params),
                 mock_clock,
                 block_time_offset_tolerance_dur, null);
+            if (blocks)
+            {
+                blocks[1 .. $].map!(b => b.header.height).each!(h => simulatePreimages(this.log, enroll_man, genesis_validator_keys, h));
+                blocks[1 .. $].each!(b => {
+                    if (this.validateBlock(b) == null)  // i.e. no invalid reason
+                        this.addValidatedBlock(b);
+                }());
+            }
```
Same thing here, feed the Ledger valid block instead of patching them inside the ctor

hewison-chris replied:
Updated.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -798,8 +798,19 @@ public class Ledger
             const CR = this.enroll_man.getCommitmentNonce(K, block.header.height);  // commited R
             if (CR == Point.init)
+            {
+                log.error("Block#{}: Validator {} (idx: {}) Couldn't find commitment for this validator",
+                          block.header.height, K, idx);
                 return "Block: Couldn't find commitment for this validator";
-            Point R = CR + challenge.toPoint();
+            }
+            const preimage = this.enroll_man.getValidatorPreimageAt(K, Height(block.header.height - 1)); // Preimage at previous block height
+            if (preimage == Hash.init)
+            {
+                log.error("Block#{}: Validator {} (idx: {}) Couldn't find pre-image for this validator",
+                          block.header.height, K, idx);
+                return "Block: Couldn't find pre-image for this validator";
+            }
```
The caller will log the error, so I don't see why we would log it here too ?

hewison-chris replied:
It will but with less information.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -387,6 +387,34 @@ public class ValidatorSet
         }
     }
 
+    /***************************************************************************
+
+        Get the utxo_key used for enrollemnt for the given public key
+
+        Params:
+            public_key = public key of the validator
+
+        Returns:
+            Return UTXO key as Hash
+
+    ***************************************************************************/
+
+    public Hash getEnrollmentForKey (in PublicKey public_key) @trusted nothrow
+    {
+        try
+        {
+            auto results = this.db.execute("SELECT key FROM validator_set " ~
+                "WHERE public_key = ? ORDER BY active DESC", public_key);
+            if (!results.empty && results.oneValue!(byte[]).length != 0)
+                return Hash(results.front.peek!(char[])(0));
+        }
+        catch (Exception ex)
+        {
+            log.error("ManagedDatabase operation error: {}", ex.msg);
+        }
+        return Hash.init;
+    }
+
```
We really ought to return all the infos about a validator instead of exposing one method per column in a row

hewison-chris replied:
I would prefer if we could have a new issue to address improving the database queries for validator information needed each block.

Geod24 replied:
Yes, that's sensible.

hewison-chris commented:
This has been rebased with latest from upstream. The tests now need to be fixed again. Will address on Monday.

hewison-chris commented:
> This has been rebased with latest from upstream. The tests now need to be fixed again. Will address on Monday.

Green again.

AndrejMitrovic commented on source/agora/consensus/data/Block.d:
```diff
@@ -551,18 +555,54 @@ public Block makeNewBlock (Transactions)(const ref Block prev_block,
+
+***************************************************************************/
+
+mixin AddLogger!();
```
This should be below the imports. The ddoc string is documenting the mixin instead of `createRandomSeed`.

hewison-chris replied:
Moved up.

AndrejMitrovic commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -360,7 +360,36 @@ public class ValidatorSet
-        Gets the number of active validators at the block height.
+        Get the utxo_key used for enrollemnt for the given public key
+
+        Params:
```
missing `height`

hewison-chris replied:
Added.

AndrejMitrovic reviewed (commented):
Just a few comments.

hewison-chris commented:
Rebased

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -112,7 +112,7 @@ public class EnrollmentManager
 
     /// Parameters for consensus-critical constants
-    private immutable(ConsensusParams) params;
+    public immutable(ConsensusParams) params;
```
I don't see it being used ?

hewison-chris replied:
> Want to submit some of those independently ? Looks nice.

Yes only last two commits are for this PR. The others are included in PRs 1658 and 1248

hewison-chris replied:
> I don't see it being used ?

Will revert, good catch, I thought I was still using it somewhere.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -591,16 +591,15 @@ public class EnrollmentManager
     ***************************************************************************/
 
-    public Hash getRandomSeed (in Hash[] keys, in Height height) @safe nothrow
+    public Hash getRandomSeed (Keys)(Keys keys, in Height height) @safe nothrow
```
You might want to use template constraints here

hewison-chris replied:
Will update in PR 1658.

hewison-chris commented:
Rebased again on latest upstream

hewison-chris committed:
Trivial replace of find with canFind

Geod24 reviewed (commented):
Want to submit some of those independently ? Looks nice.

hewison-chris commented:
To reduce the changes in this PR I created PR https://github.com/bosagora/agora/pull/2104 with the changes for random seed and well known preimage cache for unit tests.

hewison-chris commented:
Rebased and should be green. (Note the first commits are from https://github.com/bosagora/agora/pull/2071)

hewison-chris commented:
Now green but should be merged after https://github.com/bosagora/agora/pull/2071.

hewison-chris commented:
Green again :-)

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1118,15 +1130,18 @@ public class Ledger
     {
         if (utxo_key == this.enroll_man.getEnrollmentKey())
+        {
+            log.info("As this utxo key is ours we will assume preimage is revealed");
```
Idempotent functions like this should not do logging

hewison-chris replied:
Updated

Geod24 commented on source/agora/consensus/validation/Block.d:
```diff
@@ -641,6 +642,7 @@ version (unittest)
         Duration block_time_tolerance = 100.seconds,
         string file = __FILE__, size_t line = __LINE__) nothrow @safe
     {
+        Hash[] pre_images = wellKnownPreimages(prev_height + 1, genesis_validator_keys);
```
Dead code ?

hewison-chris replied:
Removed

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -360,7 +360,37 @@ public class ValidatorSet
+
+    ***************************************************************************/
+
+    public Hash getEnrollmentForKey (in Height height, in PublicKey public_key) @trusted nothrow
+    {
+        try
+        {
+            auto results = this.db.execute("SELECT key FROM validator " ~
+                "WHERE public_key = ? AND enrolled_height >= ? AND enrolled_height < ? ",
+                public_key, this.minEnrollmentHeight(height), height);
+            if (!results.empty && results.oneValue!(byte[]).length != 0)
+                return Hash(results.front.peek!(char[])(0));
+        }
+        catch (Exception ex)
+        {
+            log.error("ManagedDatabase operation error: {}", ex.msg);
+        }
+        return Hash.init;
```
This is only needed because the Ledger iterates on `PublicKey` instead of `Hash` right ?

hewison-chris replied:
Currently we use PublicKey for signatures map. This can be removed after the move to using utxo everywhere is completed.

Geod24 commented on source/agora/consensus/data/Block.d:
```diff
@@ -843,3 +847,55 @@ unittest
+    Hash random_seed2 = "seed2".hashFull();
+    Hash preimage = "preimage".hashFull();
+
+    // Generate two blocks at height 1
+    auto block1 = GenesisBlock.makeNewBlock(
+        genesisSpendable().take(2).map!(txb => txb.sign()).array(), 10, random_seed1, 6);
+    auto block2 = GenesisBlock.makeNewBlock(
+        genesisSpendable().take(3).map!(txb => txb.sign()).array(), 10, random_seed2, 6);
```
Why is the call to `array` used ? Also can we name those constants ?

hewison-chris replied:
Updated

Geod24 commented on source/agora/consensus/data/Block.d:
```diff
@@ -843,3 +847,55 @@ unittest
+    Scalar c2 = block2.hashFull();
+    assert(c1 != c2);
+
+    // Sign with same r twice
+    Signature sig1 = sign(v, R, r, c1);
+    Signature sig2 = sign(v, R, r, c2);
```
Why is this not using `createBlockSignature` ?

hewison-chris replied:
Yes good idea. Updated.

TrustHenry commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -703,6 +703,23 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    public Hash getOurPreimage (in Height height) @safe nothrow
+    {
+        return this.cycle[height];
+    }
```
https://github.com/bosagora/agora/pull/2113

Oh, you've already created this function.
I need this :)

hewison-chris replied:
Hopefully merged soon.

hewison-chris replied:
@TrustHenry: I also split this into a separate commit if you want to cherry-pick it before it is merged.

hewison-chris committed:
Add a function to return a preimage of our node

When fetching our own preimage we can go directly to the preimage cache.

hewison-chris commented:
IMO ready for merging.

hewison-chris commented on source/agora/consensus/data/Block.d:
```diff
@@ -843,3 +847,58 @@ unittest
+    assert(verify(sig1, c1, V));
+    assert(verify(sig2, c2, V));
+
+    // Calculate the private key by subtraction
+    // `s = r + (v * c)`
+    // `s1 - s2 = r + (v * c1) - (r + v * c2) = v(c1 - c2)`
+    // `v = (s1 - s2) / (c1 - c2)`
+    Scalar s = (sig1.s - sig2.s);
+    Scalar c = (c1 - c2);
+
+    Scalar secret = s * c.invert();
+    assert(secret == v,
+        format!"Key %s is not matching key %s"
+        (secret.toString(PrintMode.Clear), v.toString(PrintMode.Clear)));
```
Calculation for revealing secret key when signing multiple blocks
